### ISSUE TITLE
Add a "height" parameter for rotate_extrude to help with threading

### ISSUE THREAD

**Issue by aprotim** [openscad/openscad#5011]
If trying to build threaded parts of any kind in OpenSCAD, there's a persistent frustration that the only in-built method to extrude a shape both vertically and rotationally is to use `linear_extrude`, which keeps the extruded polygon parallel to the XY plane. Ideally, when creating threads (or similar helical structures, like corkscrews, corkscrew marble channels, etc), it would be possible to use 2D geometry to describe the vertical cross-section of the helix and extrude that along a helical path. 

**Describe the solution you'd like**
In a similar way that the `angle` parameter of `linear_extrude` extends linear (vertical) extrusion to do rotation as well, an optional `height` parameter for `rotate_extrude` could extend rotate extrusion to include vertical translation as well.

**Describe alternatives you've considered**
Common solutions involve implementing the rotation in OpenSCAD using `hull()` (which only works with convex cross-sections, or using `linear_extrude()` as in [this user-created library](https://www.thingiverse.com/thing:1958354), but these methods are painfully slow, and ad hoc for a function that is useful in many modeling contexts.

**Additional context**
I have implemented the `height` parameter in my copy of openscad (PR incoming), and the speedup for spiralizing a semi-complex (non-convex) thread cross-section is 100x in preview mode compared to the linked `spiral_extrude` library and 200x compared to my home-brewed `hull()`-based implementation (done by decomposing the cross-section into convex positive shapes and convex negative shapes, and differenceing the hulls). For a full render, the difference is even starker. The C++ implementation finishes in < 0.15s. Both other methods take 10s of minutes. (I'm still waiting for the faster one to finish, it's been at least 20 minutes. I can give actual benchmark numbers later, if required.)

**Comment by aprotim** [comment]
Ah, ok. The spiral extrude method ended up with

Total rendering time: 0:33:55.396

**Comment by pca006132** [comment]
Did you try snapshot with manifold?

**Comment by pca006132** [comment]
The `spiral_extrude` library renders in 0.5s when manifold is enabled... don't think it is so slow.

**Comment by aprotim** [comment]
Even with manifold, for a small example (attached as a .txt), spiral_extrude takes about 5x longer to render and has significant noticeable artifacts, whereas rotate_extrude renders precisely the intended geometry (for a given segmentation resolution) at a fraction of the time.

You can use the customizer on the attached script to toggle between spiral_extrude and rotate_extrude with a height parameter, if you apply my patch in your local tree.  
[screw top w spiral extrude.txt](https://github.com/openscad/openscad/files/14409270/screw.top.w.spiral.extrude.txt)

**Comment by jimb15122** [comment]
Has this been put in a nightly build yet?  It would make generating screw theads so much easier.

**Comment by jordanbrown0** [comment]
No.  https://github.com/openscad/openscad/blob/72c9919d63116f8e711f3566ae34e9eb63a2d6e6/src/core/RotateExtrudeNode.cc#L50
(You don't have to be able to read C++ to see that there's no "h" or "height" parameter specified.)

There are any number of thread libraries that will build you standard (or non-standard) threads.  Not quite as satisfying as DIY, but maybe easier.  For instance: https://github.com/BelfrySCAD/BOSL2/wiki/threading.scad

**Comment by gsohler** [comment]
Look into here:

https://imgur.com/a/AxW90oR

or

at the end of

https://pythonscad.org/tutorial/site/python_new/

**Comment by jordanbrown0** [comment]
Note also that the thread libraries will tend to do things that straightforward extrusion won't do, like tapering and blunting the ends of the thread.

**Comment by coryrc** [comment]
So there should be two parts to make this easy:

1. The plane shouldn't only be YZ, it should optionally be tilted to be perpendicular to the end of the helix. Thus polygon defines the cross-section of the thread (e.g. buttress is right triangle).
2. There should also be a rotate_extrude *function* which reports where the start and end faces are. Technically that's available if you examine the geometry, but it's not trivial to get, yet the C++ code will already exist to calculate, so provide it to the users to add the ends to the thread.

**Comment by coryrc** [comment]
And to copy some notes from my dupe:

I believe `linear_extrude` and `rotate_extrude` actually two specific implementations of a slightly-more-general extrusion, so we could probably reduce code duplication by (internally) combining their implementations.

If we allow offloading to CSG, then caching can be useful and the time taken for incremental runs can drop significantly, and even the first run since C++ is 2 OOM quicker than OpenSCAD.

**Comment by jordanbrown0** [comment]
The big question is how many times you call a thread function.  Reducing a 100ms operation to 1ms is interesting if you call it a thousand times... not nearly so interesting if you call it five times.

What does "where the start and end faces are" mean?

**Comment by coryrc** [comment]
> The big question is how many times you call a thread function.  Reducing a 100ms operation to 1ms is interesting if you call it a thousand times... not nearly so interesting if you call it five times.

It's important because it's a blocking delay which happens *every single time* you Preview or Render. It's a BIG DEAL how long an incremental compilation takes (how many man-lifetimes have been invested in bazel and caching builds? (many, if you don't know, plus hundreds of millions of dollars)). The difference between five seconds and thirty seconds can be the difference between continuing or falling out of flow state.

> What does "where the start and end faces are" mean?  

If you want to make tapered ends of the thread, you need to know exactly where the extrusion started and stopped. The code knows this usefully, then throws it away when the mesh is done. By "where [they] are" I'm not not constraining exactly what form this takes: whether this should be a unit vector matching the polygon's Z-axis but rotated for the final orientation plus a point vector representing the translation; or if just the translated points be supplied, or what.

**Comment by jordanbrown0** [comment]
I suggest that we work on performance in cases where there are demonstrated practical concerns, not based on micro benchmarks.

Before we spend a lot of time inventing such a mechanism, I'd suggest asking somebody who maintains a threading library whether, if the feature was available, they'd bother using it.  Would it make their library enough faster that their users would notice?  Would it make the library enough simpler that they'd change to reduce the complexity?  I don't know the answers, but I suspect them.

---

"Exactly where the extrusion started and stopped" seems readily calculable, and I'm not sure what form you'd return it in that would be more useful than a DIY calculation.  You could return, for instance, the list of 3D points of the initial face, but that would be a straightforward transform of the input points.  The final face is the same, but rotated and translated.  That's all the information you could possibly give them, and they'd still have to figure out where the **next** points go and that seems much harder.
